File: 2022_11_19_leetcode_in_csharp/LeetCode/P0021_MergeTwoSortedLists.cs
// https://leetcode.com/problems/merge-two-sorted-lists/

namespace LeetCode;

using static LList;

class P0021_MergeTwoSortedLists
{
    static LList<int>? MergeTwoLists(LList<int>? list1, LList<int>? list2)
    {
        return (list1, list2) switch
        {
            (var list, null) => list,
            (null, var list) => list,
            ((var head1, var tail1), (var head2, var tail2)) =>
                head1 == head2
                    ? new(head1, new(head2, MergeTwoLists(tail1, tail2)))
                    : (head1 < head2
                        ? new(head1, MergeTwoLists(tail1, list2))
                        : new(head2, MergeTwoLists(list1, tail2)))
        };
    }

    public static void Run()
    {
        var inputs = new[]
        {
            (LListFrom(1,2,4),LListFrom(1,3,4)),
            (LListFrom<int>(),LListFrom<int>()),
        };

        foreach (var (list1, list2) in inputs)
        {
            var output = MergeTwoLists(list1, list2);
            Console.WriteLine($" {list1.ToEnumerable().JoinToString()} - {list2.ToEnumerable().JoinToString()} -> {output.ToEnumerable().JoinToString()}");
        }
    }
}

File: 2022_11_03_leetcode_in_fsharp/LeetCode/P0021_MergeTwoSortedLists.fs
// url: https://leetcode.com/problems/merge-two-sorted-lists/
// tags: list, pattern-matching, rec
// examples: [ 1; 2; 4 ] [ 1; 3; 4 ] -> [1; 1; 2; 3; 4; 4]

module LeetCode.P0021_MergeTwoSortedLists

open Utils

let rec mergeTwoLists list1 list2 =
    match list1, list2 with
    | lst, [] -> lst
    | [], lst -> lst
    | head1 :: tail1, head2 :: tail2 ->
        if head1 = head2 then head1 :: head2 :: mergeTwoLists tail1 tail2
        else if head1 < head2 then head1 :: mergeTwoLists tail1 list2
        else head2 :: mergeTwoLists list1 tail2

mergeTwoLists [ 1; 2; 4 ] [ 1; 3; 4 ] === [ 1; 1; 2; 3; 4; 4 ]
mergeTwoLists [] ([]: list<int>) === []
